File: programmable_video/android/src/main/kotlin/twilio/flutter/twilio_programmable_video/BaseListener.kt
package twilio.flutter.twilio_programmable_video

import com.twilio.video.TwilioException
import io.flutter.plugin.common.EventChannel.EventSink
import java.lang.Exception

open class BaseListener {
    var events: EventSink? = null

    private fun exceptionToMap(e: Exception?): Map<String, Any?>? {
        if (e == null)
            return null

        if (e is TwilioException)
            return mapOf("code" to e.code, "message" to e.message)

        return mapOf("code" to e.message)
    }

    protected fun sendEvent(name: String, data: Any, e: Exception? = null) {
        val eventData = mapOf("name" to name, "data" to data, "error" to exceptionToMap(e))
        events?.success(eventData)
    }
}

File: programmable_video/ios/Classes/BaseListener.swift
import Flutter
import Foundation
import TwilioVideo

public class BaseListener: NSObject {
    public var events: FlutterEventSink?

    private func errorToDict(_ error: Error?) -> [String: Any]? {
        if let error = error as NSError? {
            return [
                "code": error.code,
                "message": error.description
            ]
        }
        return nil
    }

    public func sendEvent(_ name: String, data: [String: Any?]? = nil, error: Error? = nil) {
        let eventData = ["name": name, "data": data, "error": errorToDict(error)] as [String: Any?]

        if let events = events {
            events(eventData)
        }
    }
}
